### PR TITLE
simple-server

### PR DESCRIPTION
**Pull request by DanielDavidyan**

### CODE REVIEW

File: server.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+const http = require('http');
+const fs = require('fs');
+
+let content;
+let lastModified;
+const dirPath = '/Users/danieldavidyan/Documents/ofek/server/simple-server/';
+const contentPath = dirPath + 'content.txt';
+const packageJsonPath = dirPath + 'package.json';
+
+fs.readFile(contentPath, 'utf-8', (err, data) => {
```
Comment (rongoldinfeld) [inline server.ts:10]: Extract to a util folder and file
This function executes at the start of the server. What happens if i change the file 2 days after the server is up? This should be called on each request

File: server.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    if (err) {
+        console.log(err);
+    }
+    content = data.toUpperCase();
```
Comment (rongoldinfeld) [inline server.ts:14]: What happens when this file is large? You are reading the whole file to the RAM. For example if data is 20GB, your server will probably will crash. Read about streams and implement.

File: server.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+})
+
+fs.stat(packageJsonPath, (err, stats) => {
```
Comment (rongoldinfeld) [inline server.ts:17]: Extract to a util folder and file
This function executes at the start of the server. What happens if i change the file 2 days after the server is up? This should be called on each request. 
Wrap as new Promise

File: server.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    if (err) {
+        console.log(err);
+    }
+    lastModified = stats.mtime.toString();
+})
+
+http.createServer(function (req, res) {
+    let postUrl = req.url;
+    if (postUrl === '/content') {
```
Comment (rongoldinfeld) [inline server.ts:26]: Adding more requests will make this if else statement huge. Can u think of another way?

File: server.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+fs.stat(packageJsonPath, (err, stats) => {
+    if (err) {
+        console.log(err);
+    }
+    lastModified = stats.mtime.toString();
+})
+
+http.createServer(function (req, res) {
+    let postUrl = req.url;
```
Comment (rongoldinfeld) [inline server.ts:25]: You do not verify the request method. This two request will work for all methods: GET, POST, PUT…

File: server.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
+const http = require('http');
```
Comment (rongoldinfeld) [inline server.ts:1]: No types? Add types for request, variables, an so on. 
Use `import {} from …`

File: server.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+http.createServer(function (req, res) {
+    let postUrl = req.url;
+    if (postUrl === '/content') {
+        res.write(content);
+    }
+    if (postUrl === '/updateTime') {
+        res.write(lastModified);
+    }
+    res.end();
+}).listen(8080);
```
Comment (rongoldinfeld) [inline server.ts:33]: Add the ability to start server port by environment variables

File: package.json [hunk 8]
```diff
@@ -0,0 +1,15 @@
+{
+  "name": "simple-server",
+  "version": "1.0.0",
+  "description": "",
+  "main": "server.ts",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
```
Comment (rongoldinfeld) [inline package.json:7]: How do u run this server? Add scripts for server start, and remove empty test script

File: server.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
+const http = require('http');
+const fs = require('fs');
+
+let content;
+let lastModified;
+const dirPath = '/Users/danieldavidyan/Documents/ofek/server/simple-server/';
```
Comment (rongoldinfeld) [inline server.ts:6]: If i ran this server of facebook server will it work? Why?

**Review by rongoldinfeld — CHANGES_REQUESTED** [review]

File: server.ts (hunk 9 above)
Comment (DanielDavidyan) [reply server.ts:6]: No, because it is local url.

Comment (rongoldinfeld) [reply server.ts:6]: How do u solve this?

File: server.ts (hunk 7 above)
Comment (DanielDavidyan) [reply server.ts:33]: Read about node environment variables

File: package.json [hunk 10]
```diff
@@ -4,12 +4,15 @@
   "description": "",
   "main": "server.ts",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "start": "npx ts-node server.ts"
```
Comment (rongoldinfeld) [inline package.json:7]: what is npx

File: server.ts [hunk 11]
```diff
@@ -1,34 +1,14 @@
...
+import {IncomingMessage, ServerResponse} from "http";
 
-fs.stat(packageJsonPath, (err, stats) => {
-    if (err) {
-        console.log(err);
-    }
-    lastModified = stats.mtime.toString();
-})
+const http = require('http');
+const PORT = process.env.PORT || 8080;
```
Comment (rongoldinfeld) [inline server.ts:5]: port, and see how you run it

File: utilities/server.utils.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+import {ServerResponse} from "http";
+
+const contentPath = resolve(__dirname, '../content.txt');
+
+export const requestsResolver: Record<string, (res: ServerResponse) => void> = {
+    '/content': (res) => getContent(res),
+    '/updateTime': (res) => getUpdateTime(res)
+}
+
+async function getContent(res: any) {
```
Comment (rongoldinfeld) [inline utilities/server.utils.ts:13]: fix res type

File: utilities/file-system.ts [hunk 13]
```diff
@@ -0,0 +1,8 @@
+import {PathLike, promises} from "fs";
+import {resolve} from "path";
+
+const contentPath = resolve(__dirname, '../package.json');
+
+export async function statAsync(path: PathLike = contentPath): Promise<any> {
```
Comment (rongoldinfeld) [inline utilities/file-system.ts:6]: any type

File: utilities/file-system.ts [hunk 14]
```diff
@@ -0,0 +1,8 @@
...
+    return (await promises.stat(path)).mtime.toString();
```
Comment (rongoldinfeld) [inline utilities/file-system.ts:7]: use fs.stat without promises, and use new Promise()

File: server.ts [hunk 15]
```diff
@@ -1,34 +1,14 @@
...
-http.createServer(function (req, res) {
+http.createServer(function (req: IncomingMessage, res: ServerResponse) {
+    res.writeHead(200);
     let postUrl = req.url;
-    if (postUrl === '/content') {
-        res.write(content);
-    }
-    if (postUrl === '/updateTime') {
-        res.write(lastModified);
+    if (requestsResolver[<string>postUrl] && req.method === 'GET') {
```
Comment (rongoldinfeld) [inline server.ts:10]: nested requests resolver to avoid multiple ifs

File: server.ts [hunk 16]
```diff
@@ -1,34 +1,14 @@
...
-    if (err) {
-        console.log(err);
-    }
-    lastModified = stats.mtime.toString();
-})
+const http = require('http');
+const PORT = process.env.PORT || 8080;
 
-http.createServer(function (req, res) {
+http.createServer(function (req: IncomingMessage, res: ServerResponse) {
```
Comment (rongoldinfeld) [inline server.ts:7]: deal with 404-  not found

File: server.ts [hunk 17]
```diff
@@ -1,34 +1,14 @@
...
+    res.writeHead(200);
```
Comment (rongoldinfeld) [inline server.ts:8]: should we call it? what happens on bad request? not found?

**Review by rongoldinfeld — CHANGES_REQUESTED** [review]

**Review by rongoldinfeld — APPROVED** [review]

**Pull request merged**